design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,241 @@
+building (in most cases this is expected to involve light customisation of an
+existing image) by creating a controller. The baremetal-operator will create
+and use RamDiskImageRequests as required when configured to do so.
+
+## Motivation
+
+When running the ironic-python-agent (IPA) deploy ramdisk, the agent is reliant
+on information baked in to the ramdisk image or available to discover from the
+immediate environment in order to be able to contact Ironic. In most cases this
+involves DHCP supplying an IP address and default route.
+
+Some users would prefer not to run DHCP in their infrastructure, but rather use
+static IP addresses instead. In some cases, particularly when provisioning
+using virtualmedia and not using PXE, they might prefer that the network on
+which to contact Ironic is not an untagged VLAN. In many cases, servers will be
+connected to switches that use static link aggregators, and thus should set up
+NIC bonding to enable reliable two-way communication.
+
+Currently there is no way to use static IPs or configure any of the other
+options on a per-host basis because all hosts are booted using the same ramdisk
+image.
+
+### Goals
+
+- Allow each host to use its own deploy ramdisk image customised with an
+  appropriate network configuration.
+
+### Non-Goals
+
+- Allow users who expect to provision a host to specify their own custom image
+  containing IPA.
+- Use Ironic with images that don't contain IPA.
+- Implement custom provisioners not based on Ironic.
+
+## Proposal
+
+### User Stories
+
+As a data center operator, I would like to manage hosts with metal³ and deploy
+them via virtualmedia on networks where there is no DHCP-like configuration
+service.
+
+## Design Details
+
+A new Custom Resource type, `RamDiskImageRequest` will be created in the
+metal3.io group. The Spec will contain the following fields:
+
+- Name of networkData Secret
+- Architecture
+
+The Status will contain the following fields:
+
+- Image URL
+- Name + Version of networkData Secret
+- Error message
+- Ready condition
+- Error condition
+
+A default controller for this resource type will be implemented that simply
+sets a fixed URL (supplied in its own environment or command line) as the
+output. However, provision will be made for interested parties to implement
+alternate controllers to allow image customisation.
+
+A command-line switch (defaulting to off) will enable integration between the
+baremetal-operator and the image request controller.
+
+### Implementation Details/Notes/Constraints
+
+When integration is enabled, the controller portion of the baremetal-operator
+will update the `RamDiskImageRequest` with the same name and namespace as the
+`BareMetalHost`, creating it if it doesn't exist, to point to the networkData
+Secret referenced in the `BareMetalHost` Spec. The Architecture will be
+obtained from the Host's hardware profile. (Currently we don't ratify the
```

Comment:
The hardware profile isn't set via automation until after inspection, so we won't be able to use that to determine the architecture unless we start requiring the user to provide a good value. We've previously said we wanted to deprecate that user-provided input entirely, in favor of more detailed fields like `rootDeviceHints`.

We chatted offline about having a different CRD to define how to acquire an image, so that different implementations could have different recipes that accept different inputs. If we do that, architecture could go into that CRD.

Reply:
> Do we need the extra layer of complexity, though?  I think we already know we're going to need multiple ways to build images. Perhaps not _simultaneously in a single cluster_, though. Can we put the cluster-specific configuration for the image building into the settings for the controller that will respond to the RamDiskImageBuilder API, so we can keep that API simple and avoid having too much info in the BMO? And, if we take that approach, can we expand to something like what I described here later if we need to?

I'm not seeing a lot of value in creating the RamDiskImageBuilder type - unlike the PersistentVolume case it's not clear to me what data would actually be stored in that. The obvious way to select among multiple controllers is to have a label that they can filter on. Perhaps we could copy the label from the BMH to avoid adding another field to the BMH API. And you can always run different controllers in different namespaces. So I think there are paths forward if we find later that we need that extra complexity.

> bike-shed: We should call the API RamDiskImage instead of RamDiskImageRequest.

Agreed.

Reply:
> I guess eventually we'll likely want to add an Architecture field to BMH

Quite likely. E.g. ironic is capable of serving different PXE options for different architectures, but it has to know the architecture.